libia2/include/ia2.h

```diff
@@ -226,28 +228,125 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+                     "# zero eax/ebx/ecx/edx\n"                                \
+                     "xor %%eax,%%eax\n"                                       \
+                     "mov %%eax,%%ebx\n"                                       \
+                     "mov %%eax,%%ecx\n"                                       \
+                     "mov %%eax,%%edx\n"                                       \
+                     "# read old pkru\n" IA2_RDPKRU                            \
+                     "\n"                                                      \
+                     "# save pkru in r12d\n"                                   \
+                     "mov %%eax,%%r12d\n"                                      \
+                     "# write new pkru"                                        \
+                     "mov_pkru_eax " #i "\n" IA2_WRPKRU                        \
+                     "\n"                                                      \
+                     "mov ia2_stackptr_" #i                                    \
+                     "@GOTTPOFF(%%rip), %%r11\n"                               \
+                     "cmpq $0x0,%%fs:(%%r11)\n"                                \
+                     "je .Lfresh_init" #i                                      \
+                     "\n"                                                      \
+                     "mov $" #i                                                \
+                     ", %%rdi\n"                                               \
+                     "call ia2_reinit_stack_err\n"                             \
+                     ".Lfresh_init" #i                                         \
+                     ":\n"                                                     \
+                     "# store the stack addr in the stack pointer\n"           \
+                     "mov %%rbp, %%fs:(%%r11)\n"                               \
+                     "# restore old pkru\n"                                    \
+                     "mov %%r12d,%%eax\n" IA2_WRPKRU                           \
+                     "\n"                                                      \
+                     "pop %%rbp\n"                                             \
+                     :                                                         \
+                     :                                                         \
+                     : "rdi", "rax", "rbx", "rcx", "rdx", "r11", "r12");       \
+  }
+
+#define INIT_RUNTIME_COMMON(n)                                                 \
+  /* The 0th compartment is unprivileged and does not protect its memory, */   \
+  /* so declare its stack pointer in the shared object that sets up the */     \
+  /* runtime instead of using the INIT_COMPARTMENT macro for it. */            \
+  __thread void *ia2_stackptr_0;                                               \
```

Comment:
I would put this down with the definition of `init_stacks` because that's where it is used. Do you need it declared higher?

Reply:
No need to have it declared higher; I'll move it as suggested.